docs(digest): README (#2212)
ChainSafe/gossamer: dot/digest/README.md

`NextGrandpaAuthorityChange` is consumed by the
[GRANDPA service](https://pkg.go.dev/github.com/ChainSafe/gossamer/lib/grandpa#Service) to determine the block number of
the next GRANDPA authority change.

`HandleDigests` is invoked by the core service for each new block; this function reads the consensus digests from the
block and dispatches actions based on the digest contents. The digest `Handler` supports digests from the BABE
(authorship) and GRANDPA (finalisation) consensus engines - the following sections describe the messages in these
digests and the actions Gossamer takes when it receives them.

## BABE Messages

[BABE](https://wiki.polkadot.network/docs/learn-consensus#block-production-babe) is a block production algorithm that
assigns block authorship rights over periods of time referred to as "slots", which are a subdivision of an "epoch". One
of the benefits of BABE is that it is able to provide a blockchain runtime with secure, decentralized randomness.
Furthermore, BABE seeks to ensure the liveness of a blockchain by defining multiple tiers of potential block producers

> for each epoch. BABE digests may contain 1 of 3 messages, each of which are described below.

Reviewer: The original sentence was a bit confusing

```suggestion
for each epoch. BABE digests may contain one of the three message types described below.
```

### Next Epoch

This message is issued by the runtime on the first block of every epoch - it provides the BABE consensus engine with the
authority set and randomness for the _next_ epoch.

### Disabled

A message of this type will contain the ID of an authority; this authority should cease all authority functionality and
all other authorities should ignore any authority-related messages from the identified authority. _Note: Gossamer has
not implemented supported for messages of this type._

### Next Config

> Messages of this type may only be issued in the first block of an epoch. These types of messages supply configuration

Reviewer: `may` or `are`? 🤔

Author: In my mind "may" implies the specification and "are"/"does" implies a correct implementation.

> Messages of this type may only be issued in the first block of an epoch. These types of messages supply configuration

Reviewer: It's just one type of message right?

```suggestion
Messages of this type may only be issued in the first block of an epoch. This message type supply configuration
```

parameters that should be applied from the _next_ epoch onwards. The parameters in this configuration relate to how
backup authorities are selected.

## GRANDPA Messages

Provable [finality](https://wiki.polkadot.network/docs/glossary#finality) securely assures participants in a blockchain
network that a block (and the [transactions](https://docs.substrate.io/v3/getting-started/glossary/#transaction) it
contains) will not be reverted. Gossamer implements the
[GRANDPA](https://wiki.polkadot.network/docs/learn-consensus#finality-gadget-grandpa) finality protocol, which, like
BABE, uttilizes a rotating validator set. The GRANDPA protocol defines the following consensus digest messages:

### Scheduled Change

These messages contain a list of new authority IDs and a delay, specified as a number of _finalised_ blocks, after which
the change should be applied.

### Forced Change

> This message is like that for scheduled changes, however the delay is calculated using _imported_ blocks (as opposed to

Reviewer: ```suggestion
This message is similar to the scheduled changes message type, however the delay is calculated using _imported_ blocks (as opposed to
```

finalised blocks), which means that the change is valid for multiple candidate chains.

### Disabled

A message of this type will contain the ID of an authority; this authority should cease all authority functionality and
all other authorities should ignore any authority-related messages from the identified authority. _Note: Gossamer has

> not implemented supported for messages of this type._

Reviewer: ```suggestion
not implemented support for messages of this type._
```

### Pause

> This messages specifies a delay after which the current authority set should be paused.

Reviewer: ```suggestion
Messages of this type specifies a delay after which the current authority set should be paused.
```

which are messages that originate from a blockchain runtime and are consumed by consensus engines. Consensus digests are
formally described in
[Section 5.1.2 of the Polkadot Specification](https://w3f.github.io/polkadot-spec/develop/_common_consensus_structures.html).
What follows is a description of Gossamer's consensus digest `Handler`, including the consensus engines the `Handler`
supports, and the messages the `Handler` may receive & corresponding actions it can dispatch.

## Handler

The [digest `Handler`](https://pkg.go.dev/github.com/ChainSafe/gossamer/dot/digest#Handler) implements the
[`Service` interface](https://pkg.go.dev/github.com/ChainSafe/gossamer/lib/services#Service), which means that it
exposes `Start` and `Stop` functions. When a
[Gossamer `Node`](https://pkg.go.dev/github.com/ChainSafe/gossamer/dot#Node) is started, a digest `Handler` is created
with the [`NewHandler`](https://pkg.go.dev/github.com/ChainSafe/gossamer/dot/digest#NewHandler) function and started
along with the Gossamer node's other services. The digest `Handler` maintains two `goroutines`, `handleBlockImport` and

> `handleBlockFinalisation`, which handle GRANDPA changes that occur when blocks are imported and finalised,

Reviewer 2: ```suggestion
`handleBlockFinalisation`, which handle imported and finalised blocks
```

respectively - these `goroutines` listen for messages on channels provided by the
[`BlockState`](https://pkg.go.dev/github.com/ChainSafe/gossamer/dot/state#BlockState).

The digest `Handler` also exposes two public functions:
[`NextGrandpaAuthorityChange`](https://pkg.go.dev/github.com/ChainSafe/gossamer/dot/digest#Handler.NextGrandpaAuthorityChange)
and [`HandleDigests`](https://pkg.go.dev/github.com/ChainSafe/gossamer/dot/digest#Handler.HandleDigests).

> `NextGrandpaAuthorityChange` is consumed by the
> [GRANDPA service](https://pkg.go.dev/github.com/ChainSafe/gossamer/lib/grandpa#Service) and is used during the pre-vote
> and pre-commit [equivocation stages](https://medium.com/polkadot-network/polkadot-consensus-part-2-grandpa-fb1963ef6c70)
> to determine the block number of the next GRANDPA authority change.

Reviewer 2: ```suggestion
`NextGrandpaAuthorityChange` is consumed by the
[GRANDPA service](https://pkg.go.dev/github.com/ChainSafe/gossamer/lib/grandpa#Service) 
to determine the block number of the next GRANDPA authority change.
```

`HandleDigests` is invoked by the core service for each new block; this function reads the consensus digests from the

> block and dispatches actions based on the messages the digests contain. The digest `Handler` supports digests from the

Reviewer 2: ```suggestion
block and dispatches actions based on the digest contents. The digest `Handler` supports digests from the
```

BABE (authorship) and GRANDPA (finalisation) consensus engines - the following sections describe the messages in these
digests and the actions Gossamer takes when it receives them.

## BABE Messages

> [BABE](https://wiki.polkadot.network/docs/learn-consensus#block-production-babe) is a proof-of-stake block authorship

Reviewer 2: ```suggestion
[BABE](https://wiki.polkadot.network/docs/learn-consensus#block-production-babe) is a block production
```

> algorithm that randomly assigns block authorship rights over periods of time referred to as "epochs". One of the

Reviewer 2: ```suggestion
algorithm that assigns block authorship rights over periods of time referred to as "slots", which are a subdivision of an "epoch". One of the
```

BABE doesn't assign "randomly" exactly, it appears random, but it's actually not

benefits of BABE is that it is able to provide a blockchain runtime with secure, decentralized randomness. Furthermore,
BABE seeks to ensure the liveness of a blockchain by defining multiple tiers of potential block producers for each
epoch. BABE digests may contain 1 of 3 messages, each of which are described below.

### Next Epoch

This message is issued by the runtime on the first block of every epoch - it provides the BABE consensus engine with the
authority set and randomness for the _next_ epoch.

### Disabled

A message of this type will contain the ID of an authority; this authority should cease all authority functionality and
all other authorities should ignore any authority-related messages from the identified authority.

### Next Config

Messages of this type may only be issued in the first block of an epoch. These types of messages supply configuration

> parameters that should be applied from the _next_ epoch onwards. The parameters in this configuration relate to how
> backup authorities are selected.

Reviewer 2: ```suggestion
parameters that should be applied from the _next_ epoch onwards. The parameters in this configuration are `C`, which is used to determine the BABE threshold, and a `SecondarySlots` boolean, which is set to true if secondary slot block production is enabled.
```

Reviewer 2: this is kinda hard cause we don't want to describe the BABE threshold and secondary slots here (or do we?)

Author: This feels like too much detail for me. I'm worried that it will create more questions than it answers. For instance, what is "the BABE threshold"?

Author: That's a rhetorical question BTW 😅 I know (kind of 😅) what it means, but I'm worried that it won't be transparent to everyone reading this.

## GRANDPA Messages

> Deterministic [finality](https://wiki.polkadot.network/docs/glossary#finality) securely assures participants in a

Reviewer 2: ```suggestion
Provable [finality](https://wiki.polkadot.network/docs/glossary#finality) securely assures participants in a
```

blockchain network that a block (and the
[transactions](https://docs.substrate.io/v3/getting-started/glossary/#transaction) it contains) will not be reverted.
Gossamer implements the [GRANDPA](https://wiki.polkadot.network/docs/learn-consensus#finality-gadget-grandpa) finality

> protocol, which, like BABE, assigns authority privileges to sets of nodes for given periods of time. The GRANDPA

Reviewer 2: ```suggestion
protocol, which, like BABE, uttilizes a rotating validator set. The GRANDPA
```

protocol defines the following consensus digest messages:

### Scheduled Change

These messages contain a list of new authority IDs and a delay, specified as a number of _finalised_ blocks, after which
the change should be applied.

### Forced Change

This message is like that for scheduled changes, however the delay is calculated using _imported_ blocks (as opposed to
finalised blocks), which means that the change is valid for multiple candidate chains.

### Disabled

A message of this type will contain the ID of an authority; this authority should cease all authority functionality and
all other authorities should ignore any authority-related messages from the identified authority. _Note: Gossamer has

> not implemented supported for messages of this type._

Reviewer 2: we don't support it for BABE either (it's essentially a no-op) because substrate hasn't implemented it either

Reviewer 2: and by implemented, I mean that the digest exists in the chain, but substrate doesn't act upon it (so neither do we)
